### ISSUE TITLE
Issue while importing pydoop inside pySpark map function

### ISSUE THREAD

**Issue by srinivasachari** [crs4/pydoop#273]
I have a requirement to write to hdfs inside map, hence am shipping pydoop.zip dependency module to all worker nodes using sc.addPyFile options,  but when I try importing pydoop.hdfs I get below error. 

```
Py4JJavaError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.runJob.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 4 times, most recent failure: Lost task 0.3 in stage 0.0 (TID 3, PPHDPWORKR046XX.global.tesco.org): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/usr/hdp/2.5.5.0-157/spark/python/pyspark/worker.py", line 111, in main
    process()
  File "/usr/hdp/2.5.5.0-157/spark/python/pyspark/worker.py", line 106, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/usr/hdp/2.5.5.0-157/spark/python/pyspark/serializers.py", line 263, in dump_stream
    vs = list(itertools.islice(iterator, batch))
  File "/usr/hdp/2.5.5.0-157/spark/python/pyspark/rdd.py", line 1293, in takeUpToNumLeft
    yield next(iterator)
  File "<ipython-input-13-06c2725f43c3>", line 2, in func
  File "/tmp/pip-build-zvn9Um/pydoop/pydoop-1.2.0.data/platlib/pydoop/__init__.py", line 194, in <module>
  File "/tmp/pip-build-zvn9Um/pydoop/pydoop-1.2.0.data/platlib/pydoop/__init__.py", line 179, in read_properties
IOError: [Errno 20] Not a directory: '/data/1/hadoop/yarn/local/usercache/ex63/appcache/application_1516896450292_51969/container_e310_1516896450292_51969_01_000002/pydoop.zip/pydoop/pydoop.properties'
```

Steps followed to create pydoop.zip
```
pip install pydoop -t ./pydoop
cd pydoop
zip -r pydoop.zip pydoop 
```
Sample pyspark code that I am trying to use pydoop inside map.

```python
from pyspark import SparkContext, SparkConf
SparkContext.setSystemProperty('spark.executor.memory', '4g')
conf = SparkConf().setAppName("pydoop test")
sc=SparkContext(conf=conf)
sc.addPyFile("/home/ex63/pydoop/pydoop.zip")
rdd=sc.parallelize([(12,34,56,67),(34,56,87,354),(345,74,33,77), (453,56,73,56)],2 )
def func(rec):
    from pydoop import hdfs
    print (rec[0])
    print (hdfs.__file__)
    #hdfs.dump("hello", "/user/ex63/temp_{}.txt".format(rec[0]))

rdd.map(func).take(10)
```

Please help me to resolve this.

**Comment by simleo** [comment]
Hi, and thanks for reporting this.

Pydoop uses dynamic extension modules, so it's [not importable from a zip archive](https://docs.python.org/2/library/zipimport.html). It *should* be importable from an egg (also supported by Spark), but this leads to the same error as above. I have just opened issue #276 for this and hope to get to it soon. In the meantime, since you most likely don't need properties anyway, you should be able to work around the problem as follows:

1. Change `pydoop/__init__.py` as follows:

```diff
--- a/pydoop/__init__.py
+++ b/pydoop/__init__.py
@@ -179,9 +179,7 @@ def read_properties(fname):
         with open(fname) as f:
             parser.readfp(AddSectionWrapper(f))
     except IOError as e:
-        if e.errno != errno.ENOENT:
-            raise
-        return None  # compile time, prop file is not there
+        return {}
     return dict(parser.items(AddSectionWrapper.SEC_NAME))
```

2. Try the following (on a worker node, or a machine with the same configuration):

```
git clone --branch 1.2.0 https://github.com/crs4/pydoop
cd pydoop
export HADOOP_HOME=/your/hadoop/home
export JAVA_HOME=/your/java/home
python setup.py build
python setup.py bdist_egg
```

You should end up with a `pydoop-1.2.0-py2.7.egg` under `dist/`. Try passing this file to `sc.addPyFile` instead of the zip one.

**Comment by simleo** [comment]
Hi,

It looks like it's not at all straightforward to make a Python package that includes native extensions importable from an egg or other archive. However, I've just tried your pyspark sample code and it works for me if I pass the unzipped installation dir to `addFile` with `recursive` set to `True`. For instance:

```
cd /tmp
pip install pydoop -t .
```

And in the pyspark code:

```python
from pyspark import SparkContext, SparkConf

SparkContext.setSystemProperty('spark.executor.memory', '4g')
conf = SparkConf().setAppName("pydoop test")
sc = SparkContext(conf=conf)
sc.addFile("/tmp/pydoop", recursive=True)
rdd = sc.parallelize([
    (12, 34, 56, 67),
    (34, 56, 87, 354),
    (345, 74, 33, 77),
    (453, 56, 73, 56)
], 2)


def func(rec):
    import sys
    from pyspark import SparkFiles
    d = SparkFiles.get("pydoop")
    sys.path.insert(0, d)
    from pydoop import hdfs
    hdfs.dump("hello", "/user/root/temp_{}.txt".format(rec[0]))


rdd.map(func).take(10)
```

Note that you need to manually alter `sys.path`, since `addFile` does not take care of that. I'm using pyspark 2.2.1.

**Comment by srinivasachari** [comment]
Thanks Simone, The option sc.addFile with recursive is working, I have tested this code in my local VM, but unfortunately the cluster where I need this has spark 1.6.3 which does not have recursive parameter for addFile method. 

Regards,
Srinivas

**Comment by simleo** [comment]
Hi,

I believe you can still make it work with the older Spark version. Build `pydoop.zip` as in the original post, add it with `sc.addFile("/your/path/to/pydoop.zip")`, then you can unpack it on the fly in the worker's code with something like this:

```python
def func(rec):
    import sys
    import zipfile
    import tempdir
    from pyspark import SparkFiles
    zip_fn = SparkFiles.get("pydoop.zip")
    d = tempfile.mkdtemp()
    with zipfile.ZipFile(zip_fn, 'r') as zipf:
        zipf.extractall(d)
    sys.path.insert(0, d)
    from pydoop import hdfs
    hdfs.dump("hello", "/user/root/temp_{}.txt".format(rec[0]))
```

**Comment by srinivasachari** [comment]
Hi, 
Thanks for this solution, it is working perfectly fine. The package in PyPi seems very old version (py3compat and other latest fixes were not there). Any plan of uploading the latest package in PyPi? 
 
I tried cloning pydoop from git and build package using `python setup.py bdist --format=zip`
that created `pydoop-2.0a0.linux-x86_64.zip`, then I unzipped it, I see pydoop folder is under the path   `opt/anaconda2/lib/python2.7/site-packages/`. To make this avaialbe with in Map I have to manually zip pydoop folder by going into that path . Is there any alternative option to `python setup.py` to create pydoop.zip(which can be importable after unzip) ? 

Regards,
Srinivas
